### PR TITLE
Fixing trying to access metadata via config block

### LINKED ISSUE #1249: HeatExchanger model build fails
We are not able to use the following code to create a model:
```

import pyomo.environ as pe # Pyomo environment
from idaes.core import FlowsheetBlock, StateBlock
from idaes.models.unit_models import HeatExchanger
from idaes.models.unit_models.heat_exchanger import delta_temperature_amtd_callback
from idaes.models.properties import iapws95

# Create an empty flowsheet and steam property parameter block.
model = pe.ConcreteModel()
property_pack = iapws95.Iapws95ParameterBlock()
model.fs = FlowsheetBlock(dynamic=False, default_property_package=property_pack)
model.fs.properties = property_pack

# Add a Heater model to the flowsheet.
model.fs.heat_exchanger = HeatExchanger(
    delta_temperature_callback=delta_temperature_amtd_callback,
    hot_side_name="shell",
    cold_side_name="tube"
)


```
This is the traceback:
```

Traceback (most recent call last):
  File "/home/shabroz/.config/JetBrains/PyCharmCE2023.2/scratches/pyomo_test.py", line 14, in <module>
    model.fs.heat_exchanger = HeatExchanger(
  File "/home/shabroz/.virtualenvs/empc/lib/python3.9/site-packages/pyomo/core/base/block.py", line 580, in __setattr__
    self.add_component(name, val)
  File "/home/shabroz/.virtualenvs/empc/lib/python3.9/site-packages/pyomo/core/base/block.py", line 1185, in add_component
    val.construct(data)
  File "/home/shabroz/.virtualenvs/empc/lib/python3.9/site-packages/pyomo/core/base/block.py", line 2191, in construct
    self._getitem_when_not_present(_idx)
  File "/home/shabroz/.virtualenvs/empc/lib/python3.9/site-packages/pyomo/core/base/block.py", line 2111, in _getitem_when_not_present
    obj = self._rule(_block, idx)
  File "/home/shabroz/.virtualenvs/empc/lib/python3.9/site-packages/pyomo/core/base/initializer.py", line 313, in __call__
    return self._fcn(parent, idx)
  File "/home/shabroz/.virtualenvs/empc/lib/python3.9/site-packages/idaes/core/base/process_block.py", line 39, in _rule_default
    b.build()
  File "/home/shabroz/.virtualenvs/empc/lib/python3.9/site-packages/idaes/models/unit_models/heat_exchanger.py", line 439, in build
    s1_metadata = config.hot_side.property_package.get_metadata()
AttributeError: 'object' object has no attribute 'get_metadata'
```

Heat exchanger does not build if the hot side is not provided a property pack but there is a default property pack defined for the flowsheet

### PR DESCRIPTION
**Pull request by andrewlee94**
## Summary/Motivation:
@vova292 reported a bug where an error was being raised by the HX unit model when using a default property package defined at the flowsheet level. Investigation revealed that this was due to the model trying to access the properties metadata via the config block which was pointing to `useDefault`, and that this was a widespread issue across the code base.

The fix is to access the metadata either via a StateBlock or a ControlVolume (which checks for `useDefault` and replaces it with the correct reference), or to use the `_get_property_block` method to update the config block if a control volume is not used. This PR addresses many (hopefully most) of these issues.

## Changes proposed in this PR:
-
-

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by bpaul4** [comment]
@andrewlee94 is this something that we test for? If not, do we need to add some so this doesn't occur again?

**Comment by andrewlee94** [comment]
@bpaul4 Unfortunately, this is very hard to test for as it is an integration issue. The only way to test is to force developers to add a test for their model to ensure it works with a default property package.

File: idaes/models/unit_models/statejunction.py [hunk 1]
```diff
@@ -98,6 +98,9 @@ def build(self):
         # Call UnitModel.build to setup dynamics
         super(StateJunctionData, self).build()
 
+        # Check for default property package
+        self._get_property_package()
+
```
[the remark this answers is not in the document]
Comment (andrewlee94) [reply idaes/models/unit_models/statejunction.py:103]: No, as this changes the config block which we generally avoid doing. Hence, I only used this where necessary (and even then I probably could have found another way to do it, but didn't want to work throguh that mess).

**Review by dallan-keylogic — APPROVED** [review]

**Review by bpaul4 — APPROVED** [review]

**Comment by vova292** [comment]
If i try an execute the following code using the hx_bug branch i get an error:
```
import pyomo.environ as pe # Pyomo environment
from idaes.core import FlowsheetBlock, StateBlock
from idaes.models.unit_models import HeatExchangerLumpedCapacitance
from idaes.models.unit_models.heat_exchanger import delta_temperature_amtd_callback
from idaes.models.properties import iapws95

# Create an empty flowsheet and steam property parameter block.
model = pe.ConcreteModel()
property_pack = iapws95.Iapws95ParameterBlock()
model.fs = FlowsheetBlock(dynamic=False, default_property_package=property_pack)
model.fs.properties = property_pack

# Add a Heater model to the flowsheet.
model.fs.heat_exchanger = HeatExchangerLumpedCapacitance(
    delta_temperature_callback=delta_temperature_amtd_callback,
    hot_side_name="shell",
    cold_side_name="tube"
)

```
I get the following traceback:

```
Traceback (most recent call last):
  File "/home/shabroz/git/idaes-pse/idaes/core/base/process_block.py", line 41, in _rule_default
    b.build()
  File "/home/shabroz/git/idaes-pse/idaes/models/unit_models/heat_exchanger_lc.py", line 210, in build
    raise ConfigurationError(
idaes.core.util.exceptions.ConfigurationError: fs.heat_exchanger dynamic heat balance cannot be used with a steady-state flowsheet
Traceback (most recent call last):
  File "/home/shabroz/.config/JetBrains/PyCharmCE2023.2/scratches/pyomo_test.py", line 14, in <module>
    model.fs.heat_exchanger = HeatExchangerLumpedCapacitance(
  File "/home/shabroz/.virtualenvs/empc/lib/python3.9/site-packages/pyomo/core/base/block.py", line 580, in __setattr__
    self.add_component(name, val)
  File "/home/shabroz/.virtualenvs/empc/lib/python3.9/site-packages/pyomo/core/base/block.py", line 1185, in add_component
    val.construct(data)
  File "/home/shabroz/.virtualenvs/empc/lib/python3.9/site-packages/pyomo/core/base/block.py", line 2191, in construct
    self._getitem_when_not_present(_idx)
  File "/home/shabroz/.virtualenvs/empc/lib/python3.9/site-packages/pyomo/core/base/block.py", line 2111, in _getitem_when_not_present
    obj = self._rule(_block, idx)
  File "/home/shabroz/.virtualenvs/empc/lib/python3.9/site-packages/pyomo/core/base/initializer.py", line 316, in __call__
    return self._fcn(parent, idx)
  File "/home/shabroz/git/idaes-pse/idaes/core/base/process_block.py", line 41, in _rule_default
    b.build()
  File "/home/shabroz/git/idaes-pse/idaes/models/unit_models/heat_exchanger_lc.py", line 210, in build
    raise ConfigurationError(
idaes.core.util.exceptions.ConfigurationError: fs.heat_exchanger dynamic heat balance cannot be used with a steady-state flowsheet
```

**Comment by dallan-keylogic** [comment]
@vova292 Why are you trying to use the `HeatExchangerLumpedCapacitance`? That's for dynamic model simulation, and it looks like you have a steady state flowsheet. If you really want to use it, you need to set the configuration option `dynamic_heat_balance=False`. 

Now, it looks like that unit model has some issues. It shouldn't be looking to the flowsheet flag for dynamics, but the unit level flag, and it should be able to create the holdup terms by supplying the argument `has_holdup=True`.

**Comment by andrewlee94** [comment]
@vova292 @dallan-keylogic In this case, the issue is that the unit model allows the `useDefault` option for the dynamic heat balance, but never checks to see what to do if that is the case. I have a fix mostly ready, I just need to write some tests.

**Review by dallan-keylogic — APPROVED** [review]
I guess the configs here make sense given that the original heat exchanger (which this model subclasses) gets the dynamic flag for the control volumes from the unit level dynamic flag, but in the bigger picture it gets things backwards. IIRC, for a liquid/liquid heat exchanger over 80% of the holdup comes from metal, and for a gas/gas heat exchanger almost 100% of it comes from the metal.

There's nothing wrong with the code here in question, though.

**Comment by andrewlee94** [comment]
@dallan-keylogic That might actually be a better solution, but would need us to overload some base methods.

**Pull request merged**